File: 2020/qr/5.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

const IMPOSSIBLE_TAB = {
    '2': [3],
    '3': [4,5,7,8]
}

function loop(num,max,n,size,ret){
    if(n == 0){
        if(num != 0){
            return false;
        }
        for(let b of ret){
            if(b == size - 1){
                return false;
            }
        }
        return true;
    }
    for(let i=Math.min(max,num);i>0;i--){
        if(num-i>=(n-1)){
            ret[i] += 1;
            if(loop(num-i,Math.min(max,i),n-1,size,ret)){
                return true;
            } 
            ret[i] -= 1;
        }
    }
    return false;
}
let size,matrix,flag;
function judge(row,col,num){
    if(flag[num-1][0][row]){
        return false;
    }
    if(flag[num-1][1][col]){
        return false;
    }
    return true;
}

function genMatrixDFS(row,col){
    if(col == size){
        row += 1;
    }
    if(row == size){
        return true;
    }
    if(matrix[row][col] == 0){
        for(let i = 1;i<size+1;i++){
            if(judge(row,col,i)){
                matrix[row][col] = i;
                flag[i-1][0][row] = 1;
                flag[i-1][1][col] = 1;
                let ret = genMatrixDFS(row,col+1)
                if(ret){
                    return true;
                }
                matrix[row][col] = 0;
                flag[i-1][0][row] = 0;
                flag[i-1][1][col] = 0;
            }
        }
        return false;
    } else {
        return genMatrixDFS(row,col+1)
    }
}

function genMatrix(eles){
    matrix = new Array(size);
    flag = new Array(size);
    
    for(let i=0;i<size;i++){
        matrix[i] = new Array(size);
        matrix[i].fill(0);
        flag[i] = new Array(2);
        flag[i][0] = new Array(size);
        flag[i][0].fill(0);
        flag[i][1] = new Array(size);
        flag[i][1].fill(0);
    }
    let matrixRow = 0;
    for(let i=1;i<=size;i++){
        if(eles[i] > 0){
            for(let j=0;j<eles[i];j++){
                const index = matrixRow + j;
                matrix[index][index] = i;
                flag[i-1][0][index] = 1;
                flag[i-1][1][index] = 1;
            }
            matrixRow += eles[i];
        }
    }
    genMatrixDFS(0,0);
    console.log(matrix);
}

function proc(index, n, k){
    if(IMPOSSIBLE_TAB[n]){
        if(IMPOSSIBLE_TAB[n].includes(k)){
            console.log('Case #%d: k %d IMPOSSIBLE',index,k);
            return;
        }
    } else {
        if(n+1 == k || n*n -1 == k){
            console.log('Case #%d: k %d IMPOSSIBLE',index,k);
            return;
        }
    }

    const eles = new Array(n+1);
    eles.fill(0);
    const ret = loop(k,n,n,n,eles);
    if(!ret){
        console.log('Case #%d: k %d IMPOSSIBLE',index,k);
    } else {
        console.log('Case #%d: k %d POSSIBLE',index,k);
        // gen matrix
        //size = n;
        // genMatrix(eles);
        let str = '';
        for(let i = 1;i<=n;i++){
            if(eles[i] > 0){
                str += (i+' ').repeat(eles[i])
            }
        }
        console.log(str);
    }
}


let caseCnt;
let index=0;
rl.on('line', function (line) {
    if(caseCnt == null){
        index = 0;
        caseCnt = parseInt(line);
    } else if(caseCnt > index) {
        index += 1;
        let arr = line.split(' ');
        proc(index,parseInt(arr[0]),parseInt(arr[1]));
    } else {
        process.exit();
    }
});


for(let i=2;i<10;i++){
    for(let j = i;j<i*i+1;j++){
        proc(i,i,j);
    }
}

File: 2020/qr/5.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <math.h>

#define MIN(a,b) (((a)<(b))?(a):(b))
#define MAX(a,b) (((a)>(b))?(a):(b))

int findTrace(int num,int max,int n, int size,char *ret){
    int i;
    if(n == 0){
        if(num != 0){
            return 0;
        }
        for(i=0;i<size;i++){
            if(ret[i] == size - 1){
                return 0;
            } 
        }
        return 1;
    }
    for(i=MIN(max,num);i>0;i--){
        if(num-i>=(n-1)){
            ret[i-1] += 1;
            if(findTrace(num-i,MIN(max,i),n-1,size,ret)){
                return 1;
            } 
            ret[i-1] -= 1;
        }
    }
    return 0;
}

int judge(int row,int col,int size, int num,char *flag){
    if(flag[(num-1) * size * 2 + 2 * row]){
        return 0;
    }
    if(flag[(num-1) * size * 2 + 2 * col + 1]){
        return 0;
    }
    return 1;
}

int genMatrixDFS(int row,int col, int size, char *matrix, char *flag){
    int i;
    if(col == size){
        col = 0;
        row += 1;
    }
    if(row == size){
        return 1;
    }
    if(matrix[row*size+col] == 0){
        for(i = size;i>0;i--){
            if(judge(row,col,size,i,flag)){
                matrix[row*size+col] = i;
                flag[(i-1)*2*size + 2*row] = 1;
                flag[(i-1)*2*size + 2*col + 1] = 1;
                if(genMatrixDFS(row,col+1,size,matrix,flag)){
                    return 1;
                }
                matrix[row*size+col] = 0;
                flag[(i-1)*2*size + 2*row] = 0;
                flag[(i-1)*2*size + 2*col + 1] = 0;
            }
        }
        return 0;
    } else {
        return genMatrixDFS(row,col+1,size,matrix,flag);
    }
}
int n;
int a[51][51] ;
int match[51] ;
char vis[51] ;
char used[51][51] ;
int dfs(int x)
{
	int i , j ;
	for(i = 1 ; i <= n ; i ++)
	{
		if(!used[x][i] && !vis[i])
		{
			vis[i] = 1 ;
			if(match[i] == -1 || dfs(match[i]))
			{
				match[i] = x ;
				return 1 ;
			}
		}
	}
	return 0 ;
}


void genMatrix(char * eles,int size){
    int i,j;
    int matrixRow = 1;
    n = size;
    memset(used , 0 , sizeof(used));
    memset(a,0,sizeof(a));
    for(i=0;i<size;i++){
        if(eles[i] > 0){
            for(j=0;j<eles[i];j++){
                int index = matrixRow + j;
                a[index][index] = i+1;
                used[index][i+1] = 1;
            }
            matrixRow += eles[i];
        }
    }
    /*
    for(i=1;i<=n;i++){
        a[1][i] = i;
        used[i][i] = 1;
    }*/


    for(i = n ; i > 0 ; i --)
	{
		int ans = 0 ;
        int first = a[i][i];
		memset(match , -1 , sizeof(match)) ;
        match[first] = i; 
		for(j = 1 ; j <= n ; j ++)
		{
            if(j == i){
                ans ++ ;
                continue;
            }
            memset(vis , 0 , sizeof(vis)) ;
            if(dfs(j))
                ans ++ ;	
		}
		if(ans < n)
	    {
	    	printf("sorry %d %d!!\n",ans,i);
	    	break ;
		}
		for(j = 1 ; j <= n ; j ++)
		  a[i][match[j]] = j ;
		for(j = 1 ; j <= n ; j ++)
		   used[j][a[i][j]] = 1 ;	
	}

    for(i=1;i<=n;i++){
        for(j=1;j<=n;j++){
            printf("%d ",a[i][j]);
        }
         printf("\n");
    }
}

void proc(int index,int n, int k){
    int ret = 1;
    char * eles;
    if(n == 3){
        if(k != 6 && k >=4 && k <= 8){
            ret = 0;
        }
    } else {
        if(n+1 == k || n*n - 1 == k){
            ret = 0;
        }
    }
    if(ret == 0){
        printf("Case #%d: IMPOSSIBLE\n",index);
        return;
    }
    printf("Case #%d: POSSIBLE\n",index);
    eles = (char *)malloc(n);
    memset(eles,0,n);
    findTrace(k,n,n,n,eles);
    genMatrix(eles,n);
    free(eles);
}

int main(){
    int T=0;
    int i;
    setvbuf(stdout, NULL, _IONBF, 0);
    setvbuf(stdin, NULL, _IONBF, 0);
    scanf("%d",&T);
    for(i = 0;i<T;i++){
        int n,k;
        scanf("%d%d",&n,&k);
        proc(i+1,n,k);
    }
    return 0;
}